File: Android/authservice/app/src/main/java/com/huawei/agc/quickstart/auth/SettingsActivity.java
/*
 * Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.huawei.agc.quickstart.auth;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import com.huawei.agconnect.auth.AGConnectAuth;
import com.huawei.agconnect.auth.AGConnectAuthCredential;
import com.huawei.agconnect.auth.VerifyCodeResult;
import com.huawei.agconnect.auth.VerifyCodeSettings;
import com.huawei.hmf.tasks.OnFailureListener;
import com.huawei.hmf.tasks.OnSuccessListener;
import com.huawei.hmf.tasks.Task;
import com.huawei.hmf.tasks.TaskExecutors;

public class SettingsActivity extends Activity {
    private static final String TAG = SettingsActivity.class.getSimpleName();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.settings_activity);
        initView();
    }

    private void initView() {
        findViewById(R.id.layout_delete).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // delete user
                AGConnectAuth.getInstance().deleteUser()
                    .addOnSuccessListener(TaskExecutors.uiThread(), new OnSuccessListener<Void>() {
                        @Override
                        public void onSuccess(Void aVoid) {
                            startActivity(new Intent(SettingsActivity.this, LoginActivity.class));
                            SettingsActivity.this.finish();
                        }
                    }).addOnFailureListener(TaskExecutors.uiThread(), new OnFailureListener() {
                    @Override
                    public void onFailure(Exception e) {
                        Toast.makeText(SettingsActivity.this, "delete user fail:" + e.getMessage(), Toast.LENGTH_SHORT).show();
                    }
                });
            }
        });

        findViewById(R.id.layout_update_email).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                updateEmail();
            }
        });

        findViewById(R.id.layout_update_phone).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                updatePhone();
            }
        });

        findViewById(R.id.layout_update_password).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // update password
                if (AGConnectAuth.getInstance().getCurrentUser() != null) {
                    AGConnectAuth.getInstance().getCurrentUser().updatePassword("new password", "verify code", AGConnectAuthCredential.Email_Provider);
                    //or
                    //AGConnectAuth.getInstance().getCurrentUser().updatePassword("new password", "verify code", AGConnectAuthCredential.Phone_Provider);
                }
            }
        });

        findViewById(R.id.layout_reset_password).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // reset password
                AGConnectAuth.getInstance().resetPassword("email", "new password", "verify code");
                //or
                //AGConnectAuth.getInstance().resetPassword("countryCode", "phoneNumber", "new password", "verifycode");
            }
        });
    }

    private void updatePhone() {
        View view = LayoutInflater.from(this).inflate(R.layout.dialog_phone, null, false);
        final AlertDialog dialog = new AlertDialog.Builder(this).setView(view).create();
        EditText editCc = view.findViewById(R.id.et_country_code);
        EditText editAccount = view.findViewById(R.id.et_account);
        EditText editCode = view.findViewById(R.id.et_verify_code);
        Button send = view.findViewById(R.id.btn_send);
        send.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String countryCode = editCc.getText().toString();
                String phoneNumber = editAccount.getText().toString();
                //build a verify code settings
                VerifyCodeSettings settings = VerifyCodeSettings.newBuilder()
                    .action(VerifyCodeSettings.ACTION_REGISTER_LOGIN)
                    .build();
                // request verify code,and waiting for the verification code to be sent to your mobile phone
                Task<VerifyCodeResult> task = AGConnectAuth.getInstance().requestVerifyCode(countryCode, phoneNumber, settings);
                task.addOnSuccessListener(TaskExecutors.uiThread(), new OnSuccessListener<VerifyCodeResult>() {
                    @Override
                    public void onSuccess(VerifyCodeResult verifyCodeResult) {
                        Toast.makeText(SettingsActivity.this, "request verify code success", Toast.LENGTH_SHORT).show();
                    }
                }).addOnFailureListener(TaskExecutors.uiThread(), new OnFailureListener() {
                    @Override
                    public void onFailure(Exception e) {
                        Toast.makeText(SettingsActivity.this, "request verify code fail:" + e.getMessage(), Toast.LENGTH_SHORT).show();
                    }
                });
            }
        });
        Button update = view.findViewById(R.id.update);
        update.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String countryCode = editCc.getText().toString();
                String phoneNumber = editAccount.getText().toString();
                String verifyCode = editCode.getText().toString();
                if (AGConnectAuth.getInstance().getCurrentUser() != null) {
                    // update phone
                    AGConnectAuth.getInstance().getCurrentUser().updatePhone(countryCode, phoneNumber, verifyCode)
                        .addOnSuccessListener(TaskExecutors.uiThread(), new OnSuccessListener<Void>() {
                            @Override
                            public void onSuccess(Void aVoid) {
                                Log.d(TAG, "updatePhone success");
                                Toast.makeText(SettingsActivity.this, "updatePhone success", Toast.LENGTH_SHORT).show();
                            }
                        }).addOnFailureListener(TaskExecutors.uiThread(), new OnFailureListener() {
                        @Override
                        public void onFailure(Exception e) {
                            Toast.makeText(SettingsActivity.this, "updatePhone fail:" + e.getMessage(), Toast.LENGTH_SHORT).show();
                        }
                    });
                }
                dialog.dismiss();
            }
        });
        dialog.show();
    }

    private void updateEmail() {
        View view = LayoutInflater.from(this).inflate(R.layout.dialog_email, null, false);
        final AlertDialog dialog = new AlertDialog.Builder(this).setView(view).create();
        EditText editAccount = view.findViewById(R.id.et_account);
        EditText editCode = view.findViewById(R.id.et_verify_code);
        Button send = view.findViewById(R.id.btn_send);
        send.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                //build a verify code settings
                String email = editAccount.getText().toString();
                VerifyCodeSettings settings = VerifyCodeSettings.newBuilder()
                    .action(VerifyCodeSettings.ACTION_REGISTER_LOGIN)
                    .build();
                Task<VerifyCodeResult> task = AGConnectAuth.getInstance().requestVerifyCode(email, settings);
                task.addOnSuccessListener(TaskExecutors.uiThread(), new OnSuccessListener<VerifyCodeResult>() {
                    @Override
                    public void onSuccess(VerifyCodeResult verifyCodeResult) {
                        Toast.makeText(SettingsActivity.this, "request verify code success", Toast.LENGTH_SHORT).show();
                    }
                }).addOnFailureListener(TaskExecutors.uiThread(), new OnFailureListener() {
                    @Override
                    public void onFailure(Exception e) {
                        Toast.makeText(SettingsActivity.this, "request verify code fail:" + e.getMessage(), Toast.LENGTH_SHORT).show();
                    }
                });
            }
        });
        Button update = view.findViewById(R.id.update);
        update.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String email = editAccount.getText().toString();
                String verifyCode = editCode.getText().toString();
                if (AGConnectAuth.getInstance().getCurrentUser() != null) {
                    AGConnectAuth.getInstance().getCurrentUser().updateEmail(email, verifyCode)
                        .addOnSuccessListener(TaskExecutors.uiThread(), new OnSuccessListener<Void>() {
                            @Override
                            public void onSuccess(Void aVoid) {
                                Toast.makeText(SettingsActivity.this, "updateEmail success", Toast.LENGTH_SHORT).show();
                            }
                        }).addOnFailureListener(TaskExecutors.uiThread(), new OnFailureListener() {
                        @Override
                        public void onFailure(Exception e) {
                            Toast.makeText(SettingsActivity.this, "updateEmail fail:" + e.getMessage(), Toast.LENGTH_SHORT).show();
                        }
                    });
                }
                dialog.dismiss();
            }
        });
        dialog.show();
    }

}

File: Android/authservice/kotlin-app/src/main/java/com/huawei/agc/quickstart/auth/SettingsActivity.kt
/*
 * Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.huawei.agc.quickstart.auth

import android.app.Activity
import android.app.AlertDialog
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import com.huawei.agconnect.auth.AGConnectAuth
import com.huawei.agconnect.auth.AGConnectAuthCredential
import com.huawei.agconnect.auth.VerifyCodeResult
import com.huawei.agconnect.auth.VerifyCodeSettings
import com.huawei.hmf.tasks.OnFailureListener
import com.huawei.hmf.tasks.OnSuccessListener
import com.huawei.hmf.tasks.Task
import com.huawei.hmf.tasks.TaskExecutors

class SettingsActivity constructor() : Activity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.settings_activity)
        initView()
    }

    private fun initView() {
        findViewById<View>(R.id.layout_delete).setOnClickListener { // delete user
            AGConnectAuth.getInstance().deleteUser()
                    .addOnSuccessListener(TaskExecutors.uiThread(), OnSuccessListener<Void?> {
                        startActivity(Intent(this@SettingsActivity, LoginActivity::class.java))
                        finish()
                    }).addOnFailureListener(TaskExecutors.uiThread(), OnFailureListener {
                        e -> Toast.makeText(this@SettingsActivity, "delete user fail:" + e.message, Toast.LENGTH_SHORT).show() })
        }
        findViewById<View>(R.id.layout_update_email).setOnClickListener { updateEmail() }
        findViewById<View>(R.id.layout_update_phone).setOnClickListener { updatePhone() }
        findViewById<View>(R.id.layout_update_password).setOnClickListener {
            // update password
            if (AGConnectAuth.getInstance().currentUser != null) {
                AGConnectAuth.getInstance().currentUser.updatePassword("new password", "verify code", AGConnectAuthCredential.Email_Provider)
                //or
                //AGConnectAuth.getInstance().getCurrentUser().updatePassword("new password", "verify code", AGConnectAuthCredential.Phone_Provider);
            }
        }
        findViewById<View>(R.id.layout_reset_password).setOnClickListener {
            // reset password
            AGConnectAuth.getInstance().resetPassword("email", "new password", "verify code")
            //or
            //AGConnectAuth.getInstance().resetPassword("countryCode", "phoneNumber", "new password", "verifycode");
        }
    }

    private fun updatePhone() {
        val view: View = LayoutInflater.from(this).inflate(R.layout.dialog_phone, null, false)
        val dialog: AlertDialog = AlertDialog.Builder(this).setView(view).create()
        val editCc: EditText = view.findViewById(R.id.et_country_code)
        val editAccount: EditText = view.findViewById(R.id.et_account)
        val editCode: EditText = view.findViewById(R.id.et_verify_code)
        val send: Button = view.findViewById(R.id.btn_send)
        send.setOnClickListener {
            val countryCode: String = editCc.text.toString()
            val phoneNumber: String = editAccount.text.toString()
            //build a verify code settings
            val settings: VerifyCodeSettings = VerifyCodeSettings.newBuilder()
                    .action(VerifyCodeSettings.ACTION_REGISTER_LOGIN)
                    .build()
            // request verify code,and waiting for the verification code to be sent to your mobile phone
            val task: Task<VerifyCodeResult> = AGConnectAuth.getInstance().requestVerifyCode(countryCode, phoneNumber, settings)
            task.addOnSuccessListener(TaskExecutors.uiThread(), object : OnSuccessListener<VerifyCodeResult?> {
                public override fun onSuccess(verifyCodeResult: VerifyCodeResult?) {
                    Toast.makeText(this@SettingsActivity, "request verify code success", Toast.LENGTH_SHORT).show()
                }
            }).addOnFailureListener(TaskExecutors.uiThread(), object : OnFailureListener {
                public override fun onFailure(e: Exception) {
                    Toast.makeText(this@SettingsActivity, "request verify code fail:" + e.message, Toast.LENGTH_SHORT).show()
                }
            })
        }
        val update: Button = view.findViewById(R.id.update)
        update.setOnClickListener(object : View.OnClickListener {
            public override fun onClick(v: View) {
                val countryCode: String = editCc.getText().toString()
                val phoneNumber: String = editAccount.getText().toString()
                val verifyCode: String = editCode.getText().toString()
                if (AGConnectAuth.getInstance().getCurrentUser() != null) {
                    // update phone
                    AGConnectAuth.getInstance().getCurrentUser().updatePhone(countryCode, phoneNumber, verifyCode)
                            .addOnSuccessListener(TaskExecutors.uiThread(), OnSuccessListener<Void?> {
                                Log.d(TAG, "updatePhone success")
                                Toast.makeText(this@SettingsActivity, "updatePhone success", Toast.LENGTH_SHORT).show()
                            }).addOnFailureListener(TaskExecutors.uiThread(), OnFailureListener { e -> Toast.makeText(this@SettingsActivity, "updatePhone fail:" + e.message, Toast.LENGTH_SHORT).show() })
                }
                dialog.dismiss()
            }
        })
        dialog.show()
    }

    private fun updateEmail() {
        val view: View = LayoutInflater.from(this).inflate(R.layout.dialog_email, null, false)
        val dialog: AlertDialog = AlertDialog.Builder(this).setView(view).create()
        val editAccount: EditText = view.findViewById(R.id.et_account)
        val editCode: EditText = view.findViewById(R.id.et_verify_code)
        val send: Button = view.findViewById(R.id.btn_send)
        send.setOnClickListener(object : View.OnClickListener {
            public override fun onClick(v: View) {
                //build a verify code settings
                val email: String = editAccount.getText().toString()
                val settings: VerifyCodeSettings = VerifyCodeSettings.newBuilder()
                        .action(VerifyCodeSettings.ACTION_REGISTER_LOGIN)
                        .build()
                val task: Task<VerifyCodeResult> = AGConnectAuth.getInstance().requestVerifyCode(email, settings)
                task.addOnSuccessListener(TaskExecutors.uiThread(), object : OnSuccessListener<VerifyCodeResult?> {
                    public override fun onSuccess(verifyCodeResult: VerifyCodeResult?) {
                        Toast.makeText(this@SettingsActivity, "request verify code success", Toast.LENGTH_SHORT).show()
                    }
                }).addOnFailureListener(TaskExecutors.uiThread(), object : OnFailureListener {
                    public override fun onFailure(e: Exception) {
                        Toast.makeText(this@SettingsActivity, "request verify code fail:" + e.message, Toast.LENGTH_SHORT).show()
                    }
                })
            }
        })
        val update: Button = view.findViewById(R.id.update)
        update.setOnClickListener(object : View.OnClickListener {
            public override fun onClick(v: View) {
                val email: String = editAccount.getText().toString()
                val verifyCode: String = editCode.getText().toString()
                if (AGConnectAuth.getInstance().getCurrentUser() != null) {
                    AGConnectAuth.getInstance().getCurrentUser().updateEmail(email, verifyCode)
                            .addOnSuccessListener(TaskExecutors.uiThread(), object : OnSuccessListener<Void?> {
                                public override fun onSuccess(aVoid: Void?) {
                                    Toast.makeText(this@SettingsActivity, "updateEmail success", Toast.LENGTH_SHORT).show()
                                }
                            }).addOnFailureListener(TaskExecutors.uiThread(), object : OnFailureListener {
                                public override fun onFailure(e: Exception) {
                                    Toast.makeText(this@SettingsActivity, "updateEmail fail:" + e.message, Toast.LENGTH_SHORT).show()
                                }
                            })
                }
                dialog.dismiss()
            }
        })
        dialog.show()
    }

    companion object {
        private val TAG: String = SettingsActivity::class.java.getSimpleName()
    }
}